client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +789,116 @@ where
+/// f(C1), nbb(D2) -> false (handle_enactment was already performed in f(C1), we should not retract
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+#[doc(hidden)]
+pub struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Option<Block::Hash>,
+	recent_finalized_block: Option<Block::Hash>,
+}
+
+impl<Block> EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	fn new() -> Self {
+		EnactmentState { recent_best_block: None, recent_finalized_block: None }
+	}
+
+	pub fn update_and_check_if_new_enactment_is_valid<PoolApi: graph::ChainApi<Block = Block>>(
+		&mut self,
+		api: &PoolApi,
+		event: &ChainEvent<Block>,
+	) -> (bool, Option<TreeRoute<Block>>) {
+		let (new_hash, finalized) = match event {
+			ChainEvent::NewBestBlock { hash, .. } => (*hash, false),
+			ChainEvent::Finalized { hash, .. } => (*hash, true),
+		};
+
+		let best_block = match self.recent_best_block {
+			Some(recent_best_hash) => recent_best_hash,
+			None => {
+				self.recent_best_block = Some(new_hash);
+				if finalized {
+					self.recent_finalized_block = Some(new_hash);
+				}
+
+				return (true, None)
+			},
+		};
+
+		// compute actual tree route from best_block to notified block, and use it instead of
+		// tree_route provided with event
+		let tree_route =
+			api.tree_route(best_block, new_hash).unwrap().expect("tree_route exists. qed.");
+
+		log::trace!(target: "txpool", "resolve hash:{new_hash:?} finalized:{finalized:?} tree_route:{tree_route:?}, best_block:{best_block:?}, finalized_block:{:?}", self.recent_finalized_block);
+
+		if let Some(finalized_block) = self.recent_finalized_block {
+			// block was already finalized
+			if finalized_block == new_hash {
+				log::trace!(target:"txpool", "handle_enactment: block already finalized: exit 3b");
+				return (false, Some(tree_route))
+			}
+
+			// check if recently finalized block is on retracted path...
+			if tree_route.retracted().iter().any(|x| x.hash == finalized_block) {
+				log::trace!(target: "txpool", "handle_enactment: recently finalized block is on retracted path: exit 1");
+				return (false, Some(tree_route))
+			}
+		}
+
+		// If there are no enacted blocks in best_block -> hash tree_route, it means that
+		// block being finalized was already enacted. (This case also covers best_block == hash)
+		if finalized {
+			self.recent_finalized_block = Some(new_hash);
+			if tree_route.enacted().len() == 0 {
+				log::trace!(
+				target: "txpool",
+				"handle_enactment: no newly enacted blocks since recent best block: exit 2"
+				);
+				return (false, Some(tree_route))
+			}
+
+			// check if the recent best_block was retracted
+			let best_block_retracted = tree_route.retracted().iter().any(|x| x.hash == best_block);
```

Comment:
```suggestion
			let best_block_retracted = tree_route.retracted().first().map(|x| x.hash == best_block).unwrap_or_default();
```
Because the route has been computed **from** the best_block then if it is in the retracted list it should be the first